project-base/storefront/hooks/useQueryParams.ts

```diff
@@ -317,14 +317,23 @@ export const useQueryParams = () => {
 
     const pushQueries = (queries: UrlQueries, isPush?: boolean, pathnameOverride?: string) => {
         // remove queries which are not set or removed
-        const filteredQueries = getFilteredQueries(queries);
+        let filteredQueries = getFilteredQueries(queries);
+
+        const asPathname = router.asPath.split('?')[0];
+        const dynamicPageQueryKey = router.pathname.match(/\[([^\]]+)\]/)?.[1];
+        if (dynamicPageQueryKey) {
+            filteredQueries = {
+                [dynamicPageQueryKey]: pathnameOverride || asPathname,
+                ...filteredQueries,
+            };
+        }
 
         router[isPush ? 'push' : 'replace'](
             {
-                pathname: pathnameOverride || router.asPath.split('?')[0],
+                pathname: router.pathname,
                 query: filteredQueries,
             },
-            undefined,
+            getUrlFromPathnameAndQueries(pathnameOverride || asPathname, filteredQueries),
             { shallow: true },
         );
```

Comment:
After discussion it seems like we don't need to pass `filteredQueries` into those to arguments. `filteredQueries` are already passed alongside with `pathname`, so Next should already pass the queries.

This way we probably don't need this `getUrlFromPathnameAndQueries` util at all. Since on other places is it simply returning just pathname. If this idea is correct let's remove `getUrlFromPathnameAndQueries` and pass directly pathname everywhere this function is used.

Reply:
Refactored